File: stylebook/bin/commands/command.js
import { spawnSync } from 'node:child_process';
import { getConfigFile } from '../files.js';

/** Abstract class for linter command. */
class Command {
    constructor(binary, configFile) {
        this.binary = binary;
        this.configFile =
            configFile !== undefined
                ? getConfigFile(configFile)
                : null;
    }

    /**
     * Returns true if package is installed.
     *
     * @returns {boolean}
     */
    isAvailable() {
        return spawnSync(
            process.platform === 'win32' ? 'where' : 'which',
            [this.binary],
            { stdio: 'ignore' },
        ).status === 0;
    }

    /**
     * Abstract method to define specific lint command for this linter.
     *
     * @param {boolean} quiet
     * @param {path[]} targetPaths
     * @returns {string[]}
     */
    getArguments(quiet, targetPaths) {
        throw new Error(`Not implemented: ${quiet}, ${targetPaths}`);
    }

    /**
     * Run lint command for the given collection of paths.
     *
     * @param {boolean} quiet
     * @param {path[]} targetPaths
     * @returns {number}
     */
    execute(quiet, targetPaths) {
        return spawnSync(
            this.binary,
            this.getArguments(quiet, targetPaths),
            {
                stdio: 'inherit',
                shell: false,
            },
        ).status ?? 1;
    }
}

export default Command;

File: cmd/commands/command.go
package commands

import (
	"os"
	"os/exec"
	"runtime"
)

type Linter interface {
	GetBinary() string
	GetArguments(quiet bool, targetPaths []string) []string
}

type BaseCommand struct {
	Binary string
	ConfigFile string
}

func (c *BaseCommand) GetBinary() string {
	return c.Binary
}

func (c *BaseCommand) IsAvailable() bool {
	cmdName := "which"
	if runtime.GOOS == "windows" {
		cmdName = "where"
	}
	return exec.Command(cmdName, c.Binary).Run() == nil
}

func (c *BaseCommand) Execute(l Linter, quiet bool, targetPaths []string) int {
	cmd := exec.Command(l.GetBinary(), l.GetArguments(quiet, targetPaths)...)
	if !quiet {
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
	}
	if err := cmd.Run(); err != nil {
		if exitError, ok := err.(*exec.ExitError); ok {
			return exitError.ExitCode()
		}
		return 1
	}
	return 0
}
